Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp

```diff
@@ -2726,6 +2727,42 @@ auto B3IRGenerator::addArrayNewDefault(uint32_t typeIndex, ExpressionType size,
     return { };
 }
 
+auto B3IRGenerator::addArrayNewData(uint32_t typeIndex, uint32_t dataIndex, ExpressionType arraySize, ExpressionType offset, ExpressionType& result) -> PartialResult
+{
+    Wasm::TypeDefinition& arraySignature = m_info.typeSignatures[typeIndex];
+    ASSERT(arraySignature.is<ArrayType>());
+    // Get the array element type
+    Wasm::FieldType fieldType = arraySignature.as<ArrayType>()->elementType();
+
+    // Determine the array length in bytes from the element type and desired array size
+    size_t elementSize = fieldType.type.elementSize();
+
+    // Trap if (offset + (arraySize * sizeof(elementType))) > len(d$dataIndex)
+    const Segment::Ptr& segment = m_info.data[dataIndex];
+    const uint32_t segmentSizeInBytes = segment->sizeInBytes;
+    Value* arrayLengthInBytes = m_currentBlock->appendNew<Value>(m_proc, Mul, origin(), m_currentBlock->appendNew<Const32Value>(m_proc, origin(), elementSize),
+        get(arraySize));
+    Value* offsetPlusArrayLength = m_currentBlock->appendNew<Value>(m_proc, Add, origin(), arrayLengthInBytes, get(offset));
+    {
+        CheckValue* check = m_currentBlock->appendNew<CheckValue>(m_proc, Check, origin(),
+            m_currentBlock->appendNew<Value>(m_proc, Above, origin(), offsetPlusArrayLength, m_currentBlock->appendNew<Const32Value>(m_proc, origin(), segmentSizeInBytes)));
+        check->setGenerator([=, this] (CCallHelpers& jit, const B3::StackmapGenerationParams&) {
+            this->emitExceptionCheck(jit, ExceptionType::OutOfBoundsDataSegmentAccess);
+        });
+    }
+
+    // FIXME: Emit this inline.
```

Comment:
I don't think this function should be inlined since it is very complex, correct?

Reply:
Since it ultimately calls `memcpy()`, I suppose you're right.